debug healthz wait flakes

pweil- opened the pull request:
Adding some extra debugging to try and reproduce the flakes in https://github.com/openshift/origin/issues/6031.

I was unable to reproduce it locally without running a test script, pausing, and rerunning the same script in another window.  This is attempting to list the parent processes that started the openshift command that seems to be causing the failure so I can see what isn't cleaning up nicely

cc @stevekuznetsov @deads2k @php-coder

pweil- commented:
[test]

stevekuznetsov commented on hack/util.sh:
```diff
@@ -127,6 +127,21 @@ function configure_os_server {
 	echo "[INFO] To debug: export KUBECONFIG=$ADMIN_KUBECONFIG"
 }
 
+# ensure_no_openshift displays existing OS processes.  Existing processes may cause
```
If we're going to keep this in `test-cmd`, s/ensure_no_openshift/check_for_running_openshift/

pweil- commented:
@php-coder @stevekuznetsov - I wasn't planning on keeping this code, just using it to help identify the issue.  I'll clean it up if we decide to keep it.

If you guys have an easier way to reproduce this flake please let me know, looks like it didn't hit it in my initial test and I wasn't able to trigger it locally.

pweil- commented:
re[test] - c'mon flake!

stevekuznetsov commented:
I have no way of reproducing this - locally, when another instance of OpenShift is running I always see this error, but I'm not certain this is what is happening on Jenkins. I don't think it's a terrible idea to get this code in to `test-cmd` and let it sit there, so the next time we see the flake we have more information about it. We *could* sit here and kick Jenkins forever but at least the other way it's being more efficient.

pweil- commented:
> I don't think it's a terrible idea to get this code in to test-cmd

That's where I'm leaning too.  

@php-coder I'm going to clean this up a bit based on suggestions however I don't think pgrep is what I need here.  I can use pgrep to find the initial process but what I really need is the parent pid of the process.  Will incorporate the other feedback. Thanks!

stevekuznetsov commented:
Looks like you're getting all the flakes you *don't* want @pweil- ...

php-coder commented:
> I can use pgrep to find the initial process but what I really need is the parent pid of the process.

@pweil- You are right: I missed the idea of gettig parent PID.

pweil- commented:
@deads2k any objection to this?

deads2k commented:
> @deads2k any objection to this?

Just confirming: this just displays the information, it doesn't actively kill anything, right?

Do you want to check etcd and openshift for `test-integration.sh` too?

stevekuznetsov commented:
@pweil- are we still trying to put this in ?

pweil- commented:
> @pweil- are we still trying to put this in ?

Yes, I got caught up with PSP items.  Will have this updated today.

pweil- commented:
> Just confirming: this just displays the information, it doesn't actively kill anything, right?
> 
> Do you want to check etcd and openshift for test-integration.sh too?

Yes, no killing, just reporting.  Added checks for test-integration.sh.  Example output when running e2e with a pause and running test-integration.sh in another window:

```
Test test/integration -tags='integration !docker etcd' ...

Building test executable...
[INFO] Scanning for openshift related processes already up
[WARNING] Existing openshift processes were detected, this may cause unexpected failures : 
[WARNING] openshift processes detected : 
  PID  PPID CMD
 7875  7866 openshift start --master-config=/tmp/openshift-e2e/openshift.local.config/master/master-config.yaml --node-config=/tmp/openshift-e2e/openshift.local.config/node-192.168.1.125/node-config.yaml --loglevel=4
[WARNING] Parent processes : 
  PID  PPID CMD
 7866  7716 sudo env PATH=hack/../_output/local/bin/linux/amd64:/home/pweil/codebase/vim-go-workspace/bin:/usr/local/go/bin:/home/pweil/codebase/vim-go-workspace/bin:/usr/local/go/bin:/usr/lib64/qt-3.3/bin:/home/pweil/codebase/vim-go-workspace/bin:/usr/local/go/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/home/pweil/.local/bin:/home/pweil/bin:/home/pweil/codebase/gotools/bin:/home/pweil/bin:/home/pweil/codebase/openshift/src/github.com/openshift/origin/Godeps/_workspace:/home/pweil/codebase/openshift/bin:/home/pweil/codebase/openshift/src/github.com/openshift/origin/_output/local/bin/linux/amd64:/opt/etcd:/home/pweil/codebase/gotools/bin OPENSHIFT_PROFILE=web OPENSHIFT_ON_PANIC=crash openshift start --master-config=/tmp/openshift-e2e/openshift.local.config/master/master-config.yaml --node-config=/tmp/openshift-e2e/openshift.local.config/node-192.168.1.125/node-config.yaml --loglevel=4
 7716 26422 /bin/bash hack/test-end-to-end.sh
```
I added an extra layer of parent pid reporting as well, it seems to be enough to get all the way up to the shell command.

pweil- commented:
anyone have any additional comments or suggestions on this (or better ways of debugging)?

deads2k commented:
Can you add a check for the give up case too?  I think I'd like to see it was just dead or what.

pweil- commented:
> Can you add a check for the give up case too? I think I'd like to see it was just dead or what.

Added check in wait_for_url if it fails

pweil- commented:
All - is this still useful or does someone have a better approach or more comments on the code?

pweil- commented:
```
Error from server: error when creating "test/fixtures/failing-dc.yaml": deploymentconfigs "failing-dc" already exists
!!! Error in hack/../hack/../test/end-to-end/core.sh:176
	'oc create -f test/fixtures/failing-dc.yaml' exited with status 1
```
re[test]

pweil- commented:
bump, any more feedback?

stevekuznetsov commented on hack/util.sh:
```diff
@@ -96,6 +96,47 @@ function configure_os_server {
+	echo "[WARNING] Parent processes : "
+	ps ww -opid,ppid,cmd --pid="${1}"
+
+	# try to show one more level.  Wish there was a nicer way to show the forest
```
`pstree`?

pweil- replied:
will try.  This impl doesn't sit well with me but it was the best I had :-)

stevekuznetsov replied:
I don't think it's supported to limit depth with `pstree` but I don't see big downside to showing everything

liggitt replied:
```
$ pstree
-bash: pstree: command not found
```

pweil- replied:
hrm, that's right.  It wasn't available...it's all coming back now.

stevekuznetsov replied:
Damn it. Some of the tests I had written for the Bash trap refactor may use `pstree`. I can refactor.

@pweil- Bash supports recursion so you could do it this way. I've done something similar here: https://github.com/stevekuznetsov/origin/commit/d24fb54b0b1697fae40257c3ece93ad7c37cc8ff#diff-16f1c9b93f0f57f6b11f9b646fcbdaa7R141 
If you do that, be careful to correctly scope everything.

pweil- replied:
since I really only care about one more level and am formatting it so it doesn't show headers the second time I think I'm ok with it as is unless anyone feels strongly about it.

If not, anything else jumping out or can we shall we get this merged?

stevekuznetsov replied:
For getting one level, this seems fine.

smarterclayton commented:
Are we even seeing this anymore?

pweil- commented:
haven't seen it in a while and this has had meh response which I think is a good sign.  closing but keeping the branch

stevekuznetsov commented:
Seeing a lot of this on the LDAP group extended test job: https://ci.openshift.redhat.com/jenkins/job/ldap_groups_extended/

pweil- commented:
gaaaahhhh

pweil- commented:
Ok @stevekuznetsov final LGTM is in your hands then!

stevekuznetsov commented:
Specifically:
https://ci.openshift.redhat.com/jenkins/job/ldap_groups_extended/4/
https://ci.openshift.redhat.com/jenkins/job/ldap_groups_extended/5/
https://ci.openshift.redhat.com/jenkins/job/ldap_groups_extended/7/

stevekuznetsov commented on hack/util.sh:
```diff
@@ -96,6 +96,47 @@ function configure_os_server {
+	local parentsParents=$(ps h -oppid --pid="${1}")
+	parentsParents="$(echo -e "${parentsParents}" | sed -e 's/^[[:space:]]*//')"
+	if [[ -n "${parentsParents}" && "${parentsParents}" -ne "0" ]]; then
+		ps h ww -opid,ppid,cmd --pid="${parentsParents}"
```
Paging @liggitt but I believe everything on `ps` that has double-dashes is a GNU extension and not OSX compatible.

liggitt replied:
ps on OSX does not have `ww` or `--pid`. `h` has no discernable effect. instead of `cmd`, it has `command`

stevekuznetsov replied:
Since this is mostly to catch and inform us of flake conditions, let's just disable on OSX.

stevekuznetsov commented:
LGTM other than GNU extensions on `ps` for OSX compat. Could refactor to use other tools or just check for `darwin` arch and not run any of this if we're on that.

smarterclayton commented:
We can probably close this, we've updated the wait and added more debugging.
